<a href="https://colab.research.google.com/github/innsvi/python_for_ds_task/blob/main/HW_15_3_%D0%A2%D0%B5%D1%81%D1%82%D0%B8_%D0%B4%D0%BB%D1%8F_%D0%BC%D0%B0%D0%BB%D0%B8%D1%85_%D0%B2%D0%B8%D0%B1%D1%96%D1%80%D0%BE%D0%BA_%D1%82%D0%B0_%D0%BF%D1%80%D0%BE%D0%BF%D0%BE%D1%80%D1%86%D1%96%D0%B9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


**КРОК 1: Запишіть нульову та альтернативну гіпотези і визначте тип тесту**.

- Відповідно до завдання зазначимо, що історично середній час від першого заходу на сайт до покупки становив  𝜇0=123.7  хвилин

- Нульова гіпотеза: за замовчуванням ми припускаємо, що нічого не змінилось
𝐻0: 𝜇 = 123.7

- Альтернативна гіпотеза: оскільки команда підозрює, що час збільшився, то відповідно
Ha​: μ>123.7

ТИП тесту: в нашому випадку тест ПРАВОСТОРОННІЙ (односторонній), оскільки в альтернативній гіпотизі ми припускаємо збільшення часу.


**КРОК 2: Обчисліть вибіркові статистики:  𝑥¯ ,  𝑠 ,  𝑛** .

In [3]:
import numpy as np

In [4]:
# занесемо дані вибірки в масив
x = np.array([128, 135, 121, 142, 126, 151, 123])

In [5]:
# обчислимо необхідні параметри

n = x.size                          # розмір вибірки (кількість в масиві)
x_bar = x.mean()                    # середнє значення по вибірці
s = x.std(ddof=1)                   # вибіркове стандартне відхилення (s), ddof=1! (наскільки аони розкидані навколо нашого середнього)

print("n =", n)
print("x̄ =", x_bar)
print("s =", s)

n = 7
x̄ = 132.28571428571428
s = 10.980502200328935


**КРОК 3 : Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості  𝛼=0.10** .


Наразі ми знаємо:
- наша вибірка мала: n=7

- стандартне відхилення популяції невідоме

- ми використовуємо вибіркове s

Це означає: наш вибір в даному випадку - це одновибірковий t-test (Student’s t-test)

Не z-test,  бо: вибірка менше 30 та 𝜎 невідоме.

ТОБТО МАЄМО ПРАВОСТОРОННІЙ t-тест.

Для нього вирпахуємо **degrees of freedom**: Df = n - 1 = 7 - 1 = 6

In [6]:
# ЗНАЙДЕМО КРИТИЧНЕ ЗНАЧЕННЯ при 𝛼=0.10

from scipy import stats

alpha = 0.10
df = n - 1

t_critical = stats.t.ppf(1 - alpha, df)

print("df =", df)
print("Критичне t-значення =", t_critical)


df = 6
Критичне t-значення = 1.4397557472577693


In [7]:
# ЗНАХОДИМО t-СТАТИСТИКУ (на скільки стандартних помилок наше середнє більше за історичне)

mu_0 = 123.7

t_stat = (x_bar - mu_0) / (s / np.sqrt(n))

print("t-статистика =", t_stat)


t-статистика = 2.0687273144186475


In [9]:
# ЗНАХОДИМО p-value (враховуючи тест правосторонній)

p_value = 1 - stats.t.cdf(t_stat, df)

print("p-value =", p_value)


p-value = 0.042017654253682646


**КРОК 4 : Прийміть рішення, чи відхиляєте ви гіпотезу  𝐻0  використовуючи p-value**.

Виходить ми отримали наступні дані:
- df=6
- t = 2.069
- p-value=0.0420
- рівень значущості  α = 0.10
- тест правосторонній

Памятаємо, що якщо p-value ≤ α → відхиляємо 𝐻0
якщо p-value > α → не відхиляємо 𝐻0

0.042<0.10 - тобто відхиляємо нульову гіпотезу (**Отже, на рівні значущості 10% є статистично значущі докази того, що середній час до покупки після редизайну збільшився**.)
	​


**КРОК 5: Напишіть висновок: чи справді редизайн сайту подовжив час до покупки**?

Оскільки p-value = 0.042 є меншим за рівень значущості α = 0.10, ми відхиляємо нульову гіпотезу. На рівні значущості 10% отримані дані надають статистично значущі докази того, що після редизайну сайту середній час до покупки збільшився.Це означає, що редизайн, ймовірно, призвів до того, що користувачі витрачають більше часу між першим заходом на сайт і здійсненням покупки.


**КРОК 6: Чи зміниться ваше рішення при зміні рівня значущості на  𝛼=0.05 .**

У нас залишаються ті самі показники:
- df=6
- t = 2.069
- p-value=0.0420
- тест правосторонній

Але міняємо рівень значущості на  α = 0.05

Але тоді все одно : 0.042<0.05, тобто Ми ВСЕ ОДНО відхиляємо нульову гіпотезу.


**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

**КРОК 1: Сформуємо нульову та альтернативну гіпотези**.

В нас історичне значення p (нульове) = 0.23 (раніше 23% впізнавали лого компанії).

Нульова гіпотеза: H0​: p=0.23 (припускаємо, що кампанія нічого не змінила)
Альтернативна гіпотеза: Ha​: p>0.23 (після кампанії люди стали краще впізнавати лого).

Тип тесту: правосторонній тест для частки.

Рівень значущості: α=0.03.

**КРОК 2: Підготовка даних для тесту частки**.




In [10]:
# КРОК 2

successes = 311      # кількість раз впізнали лого
nobs = 1200          # розмір вибірки
p0 = 0.23            # історична частка
alpha = 0.03         # рівень значущості

# вибіркова частка (для розуміння, не для тесту)
p_hat = successes / nobs

print("Кількість успіхів (x):", successes)
print("Розмір вибірки (n):", nobs)
print("Історична частка (p0):", p0)
print("Вибіркова частка (p̂):", p_hat)


Кількість успіхів (x): 311
Розмір вибірки (n): 1200
Історична частка (p0): 0.23
Вибіркова частка (p̂): 0.25916666666666666


**Крок 3: Виконаємо z-тест для частки та перевіримо минулі результати**

In [11]:
from statsmodels.stats.proportion import proportions_ztest

In [12]:
# виконуємо z-test для пропорції
z_stat, p_value = proportions_ztest(
    count=successes,     # кількість впізнань
    nobs=nobs,           # розмір вибірки
    value=p0,            # p0 = 0.23
    alternative='larger' # Ha: p > p0 (правосторонній тест)
)

print("="*60)
print("Z-TEST ДЛЯ ПРОПОРЦІЇ (statsmodels)")
print("="*60)
print(f"z-статистика = {z_stat:.4f}")
print(f"p-value      = {p_value:.6f}")
print(f"alpha        = {alpha}")


Z-TEST ДЛЯ ПРОПОРЦІЇ (statsmodels)
z-статистика = 2.3058
p-value      = 0.010560
alpha        = 0.03


**КРОК 4: Приймаємо рішення на основі данних**

З минулого кроку бачимо:
- z-статистика = 2.3058
- p-value = 0.01056
- рівень значущості  α=0.03
- тест: правосторонній

Памятаємо
- якщо p-value ≤ α → відхиляємо H0​
- якщо p-value > α → не відхиляємо H0

У нас: *0.01056 < 0.03*, **тобто ми відхиляємо нульову гіпотезу і робимо висновок, що на рівні значущості 3% отримані дані надають статистично значущі докази того, що частка дорослих, які впізнають логотип, перевищує 23%**.